File: apple/CabalmailKit/Sources/CabalmailKit/Models/Rss.swift
import Foundation

// Wire types for the RSS reader API (`docs/rss.md`), field for field. The
// server promises additive evolution: every field here is documented, new
// fields may appear and are ignored, and fields the server has always sent
// are decoded leniently (missing string -> "", missing number -> 0) so a
// build never fails to decode a row an older or newer Lambda produced.

// MARK: - Enums

/// Per-subscription item ordering. Stored server-side, applied client-side.
public enum RssOrderingMode: String, Codable, Sendable, CaseIterable, Identifiable {
    case newestFirst = "newest_first"
    case oldestFirst = "oldest_first"
    case newestDayOldestWithin = "newest_day_oldest_within"
    case oldestDayNewestWithin = "oldest_day_newest_within"

    public var id: String { rawValue }
}

/// What opening an item shows first: the in-feed body or the publisher's page.
public enum RssOpenMode: String, Codable, Sendable, CaseIterable, Identifiable {
    case summary
    case article

    public var id: String { rawValue }
}

/// Reader-styled or the publisher's own styling, for the in-feed body and the article view.
public enum RssStyling: String, Codable, Sendable, CaseIterable, Identifiable {
    case reader
    case native

    public var id: String { rawValue }
}

/// Per-feed remote-content default. `inherit` defers to the client's global
/// remote-content preference; `show` / `hide` override it for this feed.
public enum RssRemoteContentMode: String, Codable, Sendable, CaseIterable, Identifiable {
    case inherit
    case show
    case hide

    public var id: String { rawValue }
}

/// `filter=` on `/rss_list_items`.
public enum RssItemFilter: String, Sendable, CaseIterable {
    case all
    case unread
    case favorite
}

/// `order=` on `/rss_list_items`.
public enum RssItemOrder: String, Sendable {
    case newest
    case oldest
}

/// Which items a listing or mark-all-read addresses.
public enum RssItemScope: Sendable, Hashable {
    case subscription(String)
    case folder(String)
    case all
}

extension RssItemScope {
    /// A compact string form for persistence (the local resume session, and
    /// later the server cursor): `all`, `sub:<subscriptionId>`,
    /// `folder:<folderId>`. Ids are opaque server strings that never contain
    /// a colon, but the parser splits on the first one regardless.
    public var token: String {
        switch self {
        case .all: return "all"
        case .subscription(let id): return "sub:\(id)"
        case .folder(let id): return "folder:\(id)"
        }
    }

    /// Inverse of `token`; nil for anything malformed (an unknown prefix, an
    /// empty id) so a stale or hand-edited value reads as "no scope".
    public init?(token: String) {
        if token == "all" {
            self = .all
            return
        }
        guard let colon = token.firstIndex(of: ":") else { return nil }
        let prefix = token[..<colon]
        let id = String(token[token.index(after: colon)...])
        guard !id.isEmpty else { return nil }
        switch prefix {
        case "sub": self = .subscription(id)
        case "folder": self = .folder(id)
        default: return nil
        }
    }
}

extension RssItemScope: Codable {
    public init(from decoder: Decoder) throws {
        let token = try decoder.singleValueContainer().decode(String.self)
        guard let scope = RssItemScope(token: token) else {
            throw DecodingError.dataCorrupted(.init(
                codingPath: decoder.codingPath,
                debugDescription: "Unrecognised RssItemScope token: \(token)"
            ))
        }
        self = scope
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(token)
    }
}

// MARK: - Catalog

/// The `feed` summary riding on each subscription (title, site, health).
public struct RssFeedSummary: Sendable, Codable, Hashable {
    public var feedId: String
    public var canonicalUrl: String
    public var feedType: String
    public var title: String
    public var description: String
    public var siteUrl: String
    public var itemCount: Int
    public var lastFetchedAt: String
    public var lastAttemptAt: String
    public var lastStatusCode: Int
    public var lastError: String
    public var consecutiveFailureCount: Int
    public var cadenceMinutes: Int
    public var nextFetchAt: String
    public var deadLettered: Bool

    public init(
        feedId: String, canonicalUrl: String = "", feedType: String = "", title: String = "",
        description: String = "", siteUrl: String = "", itemCount: Int = 0,
        lastFetchedAt: String = "", lastAttemptAt: String = "", lastStatusCode: Int = 0,
        lastError: String = "", consecutiveFailureCount: Int = 0, cadenceMinutes: Int = 0,
        nextFetchAt: String = "", deadLettered: Bool = false
    ) {
        self.feedId = feedId
        self.canonicalUrl = canonicalUrl
        self.feedType = feedType
        self.title = title
        self.description = description
        self.siteUrl = siteUrl
        self.itemCount = itemCount
        self.lastFetchedAt = lastFetchedAt
        self.lastAttemptAt = lastAttemptAt
        self.lastStatusCode = lastStatusCode
        self.lastError = lastError
        self.consecutiveFailureCount = consecutiveFailureCount
        self.cadenceMinutes = cadenceMinutes
        self.nextFetchAt = nextFetchAt
        self.deadLettered = deadLettered
    }

    private enum CodingKeys: String, CodingKey {
        case feedId = "feed_id", canonicalUrl = "canonical_url", feedType = "feed_type"
        case title, description, siteUrl = "site_url", itemCount = "item_count"
        case lastFetchedAt = "last_fetched_at", lastAttemptAt = "last_attempt_at"
        case lastStatusCode = "last_status_code", lastError = "last_error"
        case consecutiveFailureCount = "consecutive_failure_count"
        case cadenceMinutes = "cadence_minutes", nextFetchAt = "next_fetch_at"
        case deadLettered = "dead_lettered"
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        feedId = try container.decode(String.self, forKey: .feedId)
        canonicalUrl = try container.decodeIfPresent(String.self, forKey: .canonicalUrl) ?? ""
        feedType = try container.decodeIfPresent(String.self, forKey: .feedType) ?? ""
        title = try container.decodeIfPresent(String.self, forKey: .title) ?? ""
        description = try container.decodeIfPresent(String.self, forKey: .description) ?? ""
        siteUrl = try container.decodeIfPresent(String.self, forKey: .siteUrl) ?? ""
        itemCount = try container.decodeIfPresent(Int.self, forKey: .itemCount) ?? 0
        lastFetchedAt = try container.decodeIfPresent(String.self, forKey: .lastFetchedAt) ?? ""
        lastAttemptAt = try container.decodeIfPresent(String.self, forKey: .lastAttemptAt) ?? ""
        lastStatusCode = try container.decodeIfPresent(Int.self, forKey: .lastStatusCode) ?? 0
        lastError = try container.decodeIfPresent(String.self, forKey: .lastError) ?? ""
        consecutiveFailureCount = try container.decodeIfPresent(Int.self, forKey: .consecutiveFailureCount) ?? 0
        cadenceMinutes = try container.decodeIfPresent(Int.self, forKey: .cadenceMinutes) ?? 0
        nextFetchAt = try container.decodeIfPresent(String.self, forKey: .nextFetchAt) ?? ""
        deadLettered = try container.decodeIfPresent(Bool.self, forKey: .deadLettered) ?? false
    }
}

/// One of the caller's subscriptions. `folderId` is "" at the root.
public struct RssSubscription: Sendable, Codable, Hashable, Identifiable {
    public var subscriptionId: String
    public var feedId: String
    public var folderId: String
    public var customTitle: String
    public var orderingMode: RssOrderingMode
    public var defaultOpenMode: RssOpenMode
    public var defaultStyling: RssStyling
    public var defaultRemoteContent: RssRemoteContentMode
    public var notificationsEnabled: Bool
    public var credentialsScheme: String
    public var readWatermark: String
    public var dataStoreUuid: String
    public var createdAt: String
    public var feed: RssFeedSummary?

    public var id: String { subscriptionId }

    /// The title to show: the user's own when set, else the feed's, else its URL.
    public var displayTitle: String {
        if !customTitle.isEmpty { return customTitle }
        if let feed, !feed.title.isEmpty { return feed.title }
        return feed?.canonicalUrl ?? feedId
    }

    public init(
        subscriptionId: String, feedId: String, folderId: String = "", customTitle: String = "",
        orderingMode: RssOrderingMode = .newestFirst, defaultOpenMode: RssOpenMode = .summary,
        defaultStyling: RssStyling = .reader, defaultRemoteContent: RssRemoteContentMode = .inherit,
        notificationsEnabled: Bool = false,
        credentialsScheme: String = "", readWatermark: String = "", dataStoreUuid: String = "",
        createdAt: String = "", feed: RssFeedSummary? = nil
    ) {
        self.subscriptionId = subscriptionId
        self.feedId = feedId
        self.folderId = folderId
        self.customTitle = customTitle
        self.orderingMode = orderingMode
        self.defaultOpenMode = defaultOpenMode
        self.defaultStyling = defaultStyling
        self.defaultRemoteContent = defaultRemoteContent
        self.notificationsEnabled = notificationsEnabled
        self.credentialsScheme = credentialsScheme
        self.readWatermark = readWatermark
        self.dataStoreUuid = dataStoreUuid
        self.createdAt = createdAt
        self.feed = feed
    }

    private enum CodingKeys: String, CodingKey {
        case subscriptionId = "subscription_id", feedId = "feed_id", folderId = "folder_id"
        case customTitle = "custom_title", orderingMode = "ordering_mode"
        case defaultOpenMode = "default_open_mode", defaultStyling = "default_styling"
        case defaultRemoteContent = "default_remote_content"
        case notificationsEnabled = "notifications_enabled", credentialsScheme = "credentials_scheme"
        case readWatermark = "read_watermark", dataStoreUuid = "data_store_uuid"
        case createdAt = "created_at", feed
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        subscriptionId = try container.decode(String.self, forKey: .subscriptionId)
        feedId = try container.decode(String.self, forKey: .feedId)
        folderId = try container.decodeIfPresent(String.self, forKey: .folderId) ?? ""
        customTitle = try container.decodeIfPresent(String.self, forKey: .customTitle) ?? ""
        // Unknown enum values from a newer server fall back to the default
        // rather than failing the whole catalog decode.
        let orderingRaw = try container.decodeIfPresent(String.self, forKey: .orderingMode) ?? ""
        orderingMode = RssOrderingMode(rawValue: orderingRaw) ?? .newestFirst
        let openRaw = try container.decodeIfPresent(String.self, forKey: .defaultOpenMode) ?? ""
        defaultOpenMode = RssOpenMode(rawValue: openRaw) ?? .summary
        let stylingRaw = try container.decodeIfPresent(String.self, forKey: .defaultStyling) ?? ""
        defaultStyling = RssStyling(rawValue: stylingRaw) ?? .reader
        let remoteRaw = try container.decodeIfPresent(String.self, forKey: .defaultRemoteContent) ?? ""
        defaultRemoteContent = RssRemoteContentMode(rawValue: remoteRaw) ?? .inherit
        notificationsEnabled = try container.decodeIfPresent(Bool.self, forKey: .notificationsEnabled) ?? false
        credentialsScheme = try container.decodeIfPresent(String.self, forKey: .credentialsScheme) ?? ""
        readWatermark = try container.decodeIfPresent(String.self, forKey: .readWatermark) ?? ""
        dataStoreUuid = try container.decodeIfPresent(String.self, forKey: .dataStoreUuid) ?? ""
        createdAt = try container.decodeIfPresent(String.self, forKey: .createdAt) ?? ""
        feed = try container.decodeIfPresent(RssFeedSummary.self, forKey: .feed)
    }
}

/// A folder in the caller's hierarchy. `parentFolderId` is "" at the root.
public struct RssFolder: Sendable, Codable, Hashable, Identifiable {
    public var folderId: String
    public var parentFolderId: String
    public var name: String
    public var displayOrder: Int

    public var id: String { folderId }

    public init(folderId: String, parentFolderId: String = "", name: String, displayOrder: Int = 0) {
        self.folderId = folderId
        self.parentFolderId = parentFolderId
        self.name = name
        self.displayOrder = displayOrder
    }

    private enum CodingKeys: String, CodingKey {
        case folderId = "folder_id", parentFolderId = "parent_folder_id", name, displayOrder = "display_order"
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        folderId = try container.decode(String.self, forKey: .folderId)
        parentFolderId = try container.decodeIfPresent(String.self, forKey: .parentFolderId) ?? ""
        name = try container.decodeIfPresent(String.self, forKey: .name) ?? ""
        displayOrder = try container.decodeIfPresent(Int.self, forKey: .displayOrder) ?? 0
    }
}

/// `/rss_list_subscriptions`.
public struct RssCatalog: Sendable, Codable, Hashable {
    public var folders: [RssFolder]
    public var subscriptions: [RssSubscription]

    public init(folders: [RssFolder], subscriptions: [RssSubscription]) {
        self.folders = folders
        self.subscriptions = subscriptions
    }
}

// MARK: - Items

/// One feed item with the caller's state folded in.
public struct RssItem: Sendable, Codable, Hashable, Identifiable {
    public var feedId: String
    public var subscriptionId: String
    public var itemId: String
    /// The item's server key; opaque, passed back on state changes.
    public var sortKey: String
    public var guid: String
    public var title: String
    public var author: String
    public var url: String
    public var publishedAt: String
    public var updatedAt: String
    public var fetchedAt: String
    /// The since-sync cursor value for this item.
    public var fetchedKey: String
    public var summaryHtml: String
    public var contentHtml: String
    public var isRead: Bool
    /// Whether `isRead` is the user's own mark (server state row) rather
    /// than the subscription's read watermark. An explicit mark is exempt
    /// from the local watermark rule, so a mark-unread on an old item
    /// survives a later listing.
    public var isReadExplicit: Bool
    public var isFavorite: Bool

    /// Stable across feeds: two feeds could carry the same sort key.
    public var id: String { "\(feedId)#\(sortKey)" }

    /// The body to render: the full content when the feed delivered it,
    /// else the summary.
    public var bodyHtml: String { contentHtml.isEmpty ? summaryHtml : contentHtml }

    public init(
        feedId: String, subscriptionId: String = "", itemId: String, sortKey: String, guid: String = "",
        title: String = "", author: String = "", url: String = "", publishedAt: String = "",
        updatedAt: String = "", fetchedAt: String = "", fetchedKey: String = "",
        summaryHtml: String = "", contentHtml: String = "", isRead: Bool = false, isReadExplicit: Bool = false,
        isFavorite: Bool = false
    ) {
        self.feedId = feedId
        self.subscriptionId = subscriptionId
        self.itemId = itemId
        self.sortKey = sortKey
        self.guid = guid
        self.title = title
        self.author = author
        self.url = url
        self.publishedAt = publishedAt
        self.updatedAt = updatedAt
        self.fetchedAt = fetchedAt
        self.fetchedKey = fetchedKey
        self.summaryHtml = summaryHtml
        self.contentHtml = contentHtml
        self.isRead = isRead
        self.isReadExplicit = isReadExplicit
        self.isFavorite = isFavorite
    }

    private enum CodingKeys: String, CodingKey {
        case feedId = "feed_id", subscriptionId = "subscription_id", itemId = "item_id"
        case sortKey = "sort_key", guid, title, author, url
        case publishedAt = "published_at", updatedAt = "updated_at", fetchedAt = "fetched_at"
        case fetchedKey = "fetched_key", summaryHtml = "summary_html", contentHtml = "content_html"
        case isRead = "is_read", isReadExplicit = "is_read_explicit", isFavorite = "is_favorite"
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        feedId = try container.decode(String.self, forKey: .feedId)
        sortKey = try container.decode(String.self, forKey: .sortKey)
        subscriptionId = try container.decodeIfPresent(String.self, forKey: .subscriptionId) ?? ""
        itemId = try container.decodeIfPresent(String.self, forKey: .itemId) ?? ""
        guid = try container.decodeIfPresent(String.self, forKey: .guid) ?? ""
        title = try container.decodeIfPresent(String.self, forKey: .title) ?? ""
        author = try container.decodeIfPresent(String.self, forKey: .author) ?? ""
        url = try container.decodeIfPresent(String.self, forKey: .url) ?? ""
        publishedAt = try container.decodeIfPresent(String.self, forKey: .publishedAt) ?? ""
        updatedAt = try container.decodeIfPresent(String.self, forKey: .updatedAt) ?? ""
        fetchedAt = try container.decodeIfPresent(String.self, forKey: .fetchedAt) ?? ""
        fetchedKey = try container.decodeIfPresent(String.self, forKey: .fetchedKey) ?? ""
        summaryHtml = try container.decodeIfPresent(String.self, forKey: .summaryHtml) ?? ""
        contentHtml = try container.decodeIfPresent(String.self, forKey: .contentHtml) ?? ""
        isRead = try container.decodeIfPresent(Bool.self, forKey: .isRead) ?? false
        isReadExplicit = try container.decodeIfPresent(Bool.self, forKey: .isReadExplicit) ?? false
        isFavorite = try container.decodeIfPresent(Bool.self, forKey: .isFavorite) ?? false
    }
}

/// One row of the caller's per-item state, from the state-sync form of
/// `/rss_list_items`: the flags an item carries, keyed so they can be
/// applied to the cached copy.
public struct RssItemState: Sendable, Codable, Hashable {
    public var feedId: String
    public var sortKey: String
    public var isRead: Bool
    public var isReadExplicit: Bool
    public var isFavorite: Bool
    public var updatedAt: String

    public init(feedId: String, sortKey: String, isRead: Bool = false, isReadExplicit: Bool = false,
                isFavorite: Bool = false, updatedAt: String = "") {
        self.feedId = feedId
        self.sortKey = sortKey
        self.isRead = isRead
        self.isReadExplicit = isReadExplicit
        self.isFavorite = isFavorite
        self.updatedAt = updatedAt
    }

    private enum CodingKeys: String, CodingKey {
        case feedId = "feed_id", sortKey = "sort_key", isRead = "is_read"
        case isReadExplicit = "is_read_explicit", isFavorite = "is_favorite", updatedAt = "updated_at"
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        feedId = try container.decode(String.self, forKey: .feedId)
        sortKey = try container.decode(String.self, forKey: .sortKey)
        isRead = try container.decodeIfPresent(Bool.self, forKey: .isRead) ?? false
        isReadExplicit = try container.decodeIfPresent(Bool.self, forKey: .isReadExplicit) ?? false
        isFavorite = try container.decodeIfPresent(Bool.self, forKey: .isFavorite) ?? false
        updatedAt = try container.decodeIfPresent(String.self, forKey: .updatedAt) ?? ""
    }
}

/// A page of the state-sync form of `/rss_list_items`. `nextSince` is
/// opaque: pass it back unchanged.
public struct RssStateSyncPage: Sendable, Codable, Hashable {
    public var states: [RssItemState]
    public var nextSince: String
    public var hasMore: Bool

    public init(states: [RssItemState], nextSince: String, hasMore: Bool) {
        self.states = states
        self.nextSince = nextSince
        self.hasMore = hasMore
    }

    private enum CodingKeys: String, CodingKey {
        case states, nextSince = "next_state_since", hasMore = "has_more"
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        states = try container.decodeIfPresent([RssItemState].self, forKey: .states) ?? []
        nextSince = try container.decodeIfPresent(String.self, forKey: .nextSince) ?? ""
        hasMore = try container.decodeIfPresent(Bool.self, forKey: .hasMore) ?? false
    }
}

/// A page of the merged listing form of `/rss_list_items`.
public struct RssItemsPage: Sendable, Codable, Hashable {
    public var items: [RssItem]
    public var nextCursor: String?

    public init(items: [RssItem], nextCursor: String? = nil) {
        self.items = items
        self.nextCursor = nextCursor
    }

    private enum CodingKeys: String, CodingKey {
        case items, nextCursor = "next_cursor"
    }
}

/// A page of the since-sync form of `/rss_list_items`.
public struct RssSyncPage: Sendable, Codable, Hashable {
    public var items: [RssItem]
    public var nextSince: String
    public var hasMore: Bool

    public init(items: [RssItem], nextSince: String, hasMore: Bool) {
        self.items = items
        self.nextSince = nextSince
        self.hasMore = hasMore
    }

    private enum CodingKeys: String, CodingKey {
        case items, nextSince = "next_since", hasMore = "has_more"
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        items = try container.decodeIfPresent([RssItem].self, forKey: .items) ?? []
        nextSince = try container.decodeIfPresent(String.self, forKey: .nextSince) ?? ""
        hasMore = try container.decodeIfPresent(Bool.self, forKey: .hasMore) ?? false
    }
}

File: android/kit/src/main/kotlin/com/cabalmail/kit/models/Rss.kt
package com.cabalmail.kit.models

import com.cabalmail.kit.settings.WireEnum
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

// Wire models for the RSS reader API (`docs/rss.md`), transcribed field for
// field from the deployed contract; the Apple `CabalmailKit` `Rss.swift` is
// the reference consumer. Every field has a default so a missing key
// decodes, and enums decode leniently through [RssWire.json]: an unknown
// value (a mode a newer server added) reads as the field's default rather
// than failing the whole response.

/** The JSON configuration every RSS payload is decoded with. */
object RssWire {
    val json =
        Json {
            ignoreUnknownKeys = true
            coerceInputValues = true
        }
}

@Serializable
enum class RssOrderingMode(
    override val wire: String,
) : WireEnum {
    @SerialName("newest_first")
    NEWEST_FIRST("newest_first"),

    @SerialName("oldest_first")
    OLDEST_FIRST("oldest_first"),

    @SerialName("newest_day_oldest_within")
    NEWEST_DAY_OLDEST_WITHIN("newest_day_oldest_within"),

    @SerialName("oldest_day_newest_within")
    OLDEST_DAY_NEWEST_WITHIN("oldest_day_newest_within"),
}

/** What opens first for an item: the feed's own content or the article. */
@Serializable
enum class RssOpenMode(
    override val wire: String,
) : WireEnum {
    @SerialName("summary")
    SUMMARY("summary"),

    @SerialName("article")
    ARTICLE("article"),
}

@Serializable
enum class RssStyling(
    override val wire: String,
) : WireEnum {
    @SerialName("reader")
    READER("reader"),

    @SerialName("native")
    NATIVE("native"),
}

/** `INHERIT` defers to the app's remote-content preference; the others override it for one feed. */
@Serializable
enum class RssRemoteContentMode(
    override val wire: String,
) : WireEnum {
    @SerialName("inherit")
    INHERIT("inherit"),

    @SerialName("show")
    SHOW("show"),

    @SerialName("hide")
    HIDE("hide"),
}

/** The item list's filter pill; feeds and feed folders open on [UNREAD] until the user picks another. */
@Serializable
enum class RssItemFilter(
    override val wire: String,
) : WireEnum {
    @SerialName("all")
    ALL("all"),

    @SerialName("unread")
    UNREAD("unread"),

    @SerialName("favorite")
    FAVORITE("favorite"),

    ;

    companion object {
        val DEFAULT_FOR_FEEDS = UNREAD
    }
}

/** The server-side sort of a merged listing (request only). */
enum class RssItemOrder(
    val wire: String,
) {
    NEWEST("newest"),
    OLDEST("oldest"),
}

/**
 * What an item list covers. The [token] form (`all` | `sub:<id>` |
 * `folder:<id>`) is the identity the session record and the cross-device
 * cursor store, the same strings the Apple client's `RssItemScope` uses, so
 * a position hands between platforms without translation.
 */
sealed class RssItemScope {
    data object All : RssItemScope()

    data class Folder(
        val folderId: String,
    ) : RssItemScope()

    data class Subscription(
        val subscriptionId: String,
    ) : RssItemScope()

    val token: String
        get() =
            when (this) {
                All -> "all"
                is Folder -> "folder:$folderId"
                is Subscription -> "sub:$subscriptionId"
            }

    companion object {
        /** Null for an unknown prefix or an empty id. */
        fun fromToken(token: String): RssItemScope? {
            if (token == "all") return All
            val separator = token.indexOf(':')
            if (separator <= 0) return null
            val id = token.substring(separator + 1)
            if (id.isEmpty()) return null
            return when (token.substring(0, separator)) {
                "sub" -> Subscription(id)
                "folder" -> Folder(id)
                else -> null
            }
        }
    }
}

/** The `feed` summary on every subscription; the health fields drive the sidebar badge. */
@Serializable
data class RssFeedSummary(
    @SerialName("feed_id") val feedId: String,
    @SerialName("canonical_url") val canonicalUrl: String = "",
    @SerialName("feed_type") val feedType: String = "",
    val title: String = "",
    val description: String = "",
    @SerialName("site_url") val siteUrl: String = "",
    @SerialName("item_count") val itemCount: Int = 0,
    @SerialName("last_fetched_at") val lastFetchedAt: String = "",
    @SerialName("last_attempt_at") val lastAttemptAt: String = "",
    @SerialName("last_status_code") val lastStatusCode: Int = 0,
    @SerialName("last_error") val lastError: String = "",
    @SerialName("consecutive_failure_count") val consecutiveFailureCount: Int = 0,
    @SerialName("cadence_minutes") val cadenceMinutes: Int = 0,
    @SerialName("next_fetch_at") val nextFetchAt: String = "",
    @SerialName("dead_lettered") val deadLettered: Boolean = false,
)

@Serializable
data class RssSubscription(
    @SerialName("subscription_id") val subscriptionId: String,
    @SerialName("feed_id") val feedId: String,
    /** Empty = root. */
    @SerialName("folder_id") val folderId: String = "",
    @SerialName("custom_title") val customTitle: String = "",
    @SerialName("ordering_mode") val orderingMode: RssOrderingMode = RssOrderingMode.NEWEST_FIRST,
    @SerialName("default_open_mode") val defaultOpenMode: RssOpenMode = RssOpenMode.SUMMARY,
    @SerialName("default_styling") val defaultStyling: RssStyling = RssStyling.READER,
    @SerialName("default_remote_content") val defaultRemoteContent: RssRemoteContentMode =
        RssRemoteContentMode.INHERIT,
    @SerialName("default_filter") val defaultFilter: RssItemFilter = RssItemFilter.DEFAULT_FOR_FEEDS,
    @SerialName("notifications_enabled") val notificationsEnabled: Boolean = false,
    @SerialName("credentials_scheme") val credentialsScheme: String = "",
    @SerialName("read_watermark") val readWatermark: String = "",
    /** The per-subscription identifier the isolated web-view profile is keyed on. */
    @SerialName("data_store_uuid") val dataStoreUuid: String = "",
    @SerialName("created_at") val createdAt: String = "",
    val feed: RssFeedSummary? = null,
) {
    /** The custom title, else the feed's, else its address, else the id. */
    val displayTitle: String
        get() =
            customTitle.ifEmpty {
                feed?.title?.ifEmpty { null } ?: feed?.canonicalUrl?.ifEmpty { null } ?: feedId
            }

    /** The row after [update], the way the server will report it. */
    fun applying(update: RssSubscriptionUpdate): RssSubscription =
        copy(
            customTitle = update.customTitle ?: customTitle,
            folderId = update.folderId ?: folderId,
            orderingMode = update.orderingMode ?: orderingMode,
            defaultOpenMode = update.defaultOpenMode ?: defaultOpenMode,
            defaultStyling = update.defaultStyling ?: defaultStyling,
            defaultRemoteContent = update.defaultRemoteContent ?: defaultRemoteContent,
            defaultFilter = update.defaultFilter ?: defaultFilter,
            notificationsEnabled = update.notificationsEnabled ?: notificationsEnabled,
        )
}

@Serializable
data class RssFolder(
    @SerialName("folder_id") val folderId: String,
    /** Empty = root. */
    @SerialName("parent_folder_id") val parentFolderId: String = "",
    val name: String = "",
    @SerialName("display_order") val displayOrder: Int = 0,
    @SerialName("default_filter") val defaultFilter: RssItemFilter = RssItemFilter.DEFAULT_FOR_FEEDS,
) {
    fun applying(update: RssFolderUpdate): RssFolder =
        copy(
            name = update.name ?: name,
            parentFolderId = update.parentFolderId ?: parentFolderId,
            displayOrder = update.displayOrder ?: displayOrder,
            defaultFilter = update.defaultFilter ?: defaultFilter,
        )
}

/** `/rss_list_subscriptions`. */
@Serializable
data class RssCatalog(
    val folders: List<RssFolder> = emptyList(),
    val subscriptions: List<RssSubscription> = emptyList(),
)

@Serializable
data class RssItem(
    @SerialName("feed_id") val feedId: String,
    /** The item's key; opaque, passed back as given. */
    @SerialName("sort_key") val sortKey: String,
    @SerialName("subscription_id") val subscriptionId: String = "",
    @SerialName("item_id") val itemId: String = "",
    val guid: String = "",
    val title: String = "",
    val author: String = "",
    val url: String = "",
    @SerialName("published_at") val publishedAt: String = "",
    @SerialName("updated_at") val updatedAt: String = "",
    @SerialName("fetched_at") val fetchedAt: String = "",
    /** The since-sync cursor value. */
    @SerialName("fetched_key") val fetchedKey: String = "",
    @SerialName("summary_html") val summaryHtml: String = "",
    @SerialName("content_html") val contentHtml: String = "",
    @SerialName("is_read") val isRead: Boolean = false,
    /** Whether [isRead] is the user's own mark rather than the watermark rule. */
    @SerialName("is_read_explicit") val isReadExplicit: Boolean = false,
    @SerialName("is_favorite") val isFavorite: Boolean = false,
) {
    /** Stable across feeds; `item_id` alone is not unique. */
    val id: String
        get() = "$feedId#$sortKey"

    val bodyHtml: String
        get() = contentHtml.ifEmpty { summaryHtml }
}

/** One row of the state-sync form of `/rss_list_items`. */
@Serializable
data class RssItemState(
    @SerialName("feed_id") val feedId: String,
    @SerialName("sort_key") val sortKey: String,
    @SerialName("item_id") val itemId: String = "",
    @SerialName("is_read") val isRead: Boolean = false,
    @SerialName("is_read_explicit") val isReadExplicit: Boolean = false,
    @SerialName("is_favorite") val isFavorite: Boolean = false,
    @SerialName("updated_at") val updatedAt: String = "",
)

/** A merged listing page; a null [nextCursor] means the server has no more. */
@Serializable
data class RssItemsPage(
    val items: List<RssItem> = emptyList(),
    @SerialName("next_cursor") val nextCursor: String? = null,
)

/** A since-sync page (ingest-time cursor). */
@Serializable
data class RssSyncPage(
    val items: List<RssItem> = emptyList(),
    @SerialName("next_since") val nextSince: String = "",
    @SerialName("has_more") val hasMore: Boolean = false,
)

/** A state-sync page. */
@Serializable
data class RssStateSyncPage(
    val states: List<RssItemState> = emptyList(),
    @SerialName("next_state_since") val nextSince: String = "",
    @SerialName("has_more") val hasMore: Boolean = false,
)

@Serializable
data class RssSubscribeResult(
    val subscription: RssSubscription,
    /** True when the caller already followed this feed. */
    val existing: Boolean = false,
)

@Serializable
data class RssUnsubscribeResult(
    @SerialName("subscription_id") val subscriptionId: String = "",
    @SerialName("feed_id") val feedId: String = "",
    @SerialName("feed_purged") val feedPurged: Boolean = false,
)

/** The optional fields of `/rss_update_subscription`; null means "leave alone". */
data class RssSubscriptionUpdate(
    val customTitle: String? = null,
    /** Empty moves the subscription to the root. */
    val folderId: String? = null,
    val orderingMode: RssOrderingMode? = null,
    val defaultOpenMode: RssOpenMode? = null,
    val defaultStyling: RssStyling? = null,
    val defaultRemoteContent: RssRemoteContentMode? = null,
    val defaultFilter: RssItemFilter? = null,
    val notificationsEnabled: Boolean? = null,
) {
    val isEmpty: Boolean
        get() =
            listOf(
                customTitle,
                folderId,
                orderingMode,
                defaultOpenMode,
                defaultStyling,
                defaultRemoteContent,
                defaultFilter,
                notificationsEnabled,
            ).all { it == null }
}

/** The optional fields of `/rss_update_folder`; null means "leave alone". */
data class RssFolderUpdate(
    val name: String? = null,
    /** Empty moves the folder to the root. */
    val parentFolderId: String? = null,
    val displayOrder: Int? = null,
    val defaultFilter: RssItemFilter? = null,
) {
    val isEmpty: Boolean
        get() = name == null && parentFolderId == null && displayOrder == null && defaultFilter == null
}

@Serializable
data class RssFolderDeleteResult(
    @SerialName("folder_id") val folderId: String = "",
    @SerialName("moved_subscriptions") val movedSubscriptions: Int = 0,
    @SerialName("moved_folders") val movedFolders: Int = 0,
    @SerialName("parent_folder_id") val parentFolderId: String = "",
)

/** One entry of `/rss_set_item_state`; a null flag is not sent. */
data class RssItemStateChange(
    val feedId: String,
    val sortKey: String,
    val isRead: Boolean? = null,
    val isFavorite: Boolean? = null,
)

@Serializable
data class RssMarkAllReadResult(
    val subscriptions: Int = 0,
    val flipped: Int = 0,
    @SerialName("read_watermark") val readWatermark: String = "",
)

@Serializable
data class RssOpmlImportFailure(
    val url: String = "",
    val code: String = "",
    @SerialName("Error") val message: String = "",
)

@Serializable
data class RssOpmlImportResult(
    val created: Int = 0,
    val existing: Int = 0,
    @SerialName("folders_created") val foldersCreated: Int = 0,
    val failed: List<RssOpmlImportFailure> = emptyList(),
)

@Serializable
data class RssOpmlExport(
    val opml: String = "",
    val filename: String = "",
)
